compiler/options.go

```diff
@@ -0,0 +1,133 @@
+package compiler
+
+import (
+	"github.com/alecthomas/protobuf/parser"
+	pb "google.golang.org/protobuf/types/descriptorpb"
+)
+
+func newFileOptions(po []*parser.Option) *pb.FileOptions {
+	if len(po) == 0 {
+		return nil
+	}
+	opts := &pb.FileOptions{}
+	for _, o := range po {
+		if len(o.Name) != 1 {
+			opts.UninterpretedOption = append(opts.UninterpretedOption, newUninterpretedOption(o))
+			continue
+		}
+		name := o.Name[0].Name
+		switch name {
+		case "java_package":
+			opts.JavaPackage = o.Value.String
+		case "java_outer_classname":
+			opts.JavaOuterClassname = o.Value.String
+		case "go_package":
+			opts.GoPackage = o.Value.String
+		case "objc_class_prefix":
+			opts.ObjcClassPrefix = o.Value.String
+		case "csharp_namespace":
+			opts.CsharpNamespace = o.Value.String
+		case "swift_prefix":
+			opts.SwiftPrefix = o.Value.String
+		case "php_class_prefix":
+			opts.PhpClassPrefix = o.Value.String
+		case "php_namespace":
+			opts.PhpNamespace = o.Value.String
+		case "php_metadata_namespace":
+			opts.PhpMetadataNamespace = o.Value.String
+		case "ruby_package":
+			opts.RubyPackage = o.Value.String
+		case "java_multiple_files":
+			opts.JavaMultipleFiles = (*bool)(o.Value.Bool)
+		case "java_generate_equals_and_hash":
+			opts.JavaGenerateEqualsAndHash = (*bool)(o.Value.Bool) // nolint:staticcheck // SA1019: opts.JavaGenerateEqualsAndHash is deprecated: Do not use. - we want to try and be maximum backwards compatible.
+		case "java_string_check_utf8":
+			opts.JavaStringCheckUtf8 = (*bool)(o.Value.Bool)
+		case "cc_generic_services":
+			opts.CcGenericServices = (*bool)(o.Value.Bool)
+		case "java_generic_services":
+			opts.JavaGenericServices = (*bool)(o.Value.Bool)
+		case "py_generic_services":
+			opts.PyGenericServices = (*bool)(o.Value.Bool)
+		case "php_generic_services":
+			opts.PhpGenericServices = (*bool)(o.Value.Bool)
+		case "deprecated":
+			opts.Deprecated = (*bool)(o.Value.Bool)
+		case "cc_enable_arenas":
+			opts.CcEnableArenas = (*bool)(o.Value.Bool)
+		case "optimize_for":
```

Comment:
Man it sucks that these are all special cased by proto :(

Reply:
yeah, now that you say it like that  I'm thinking: maybe I went about it the wrong way and should have just directly unmarshalled into the corresponding Message|Field|Service...Options.
I might end up refactoring it once I've got the custom options sorted, @alecthomas